streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -377,8 +370,9 @@ private void handleCloseAndRecycle(final Set<Task> tasksToRecycle,
         for (final Task task : tasksToCloseClean) {
             try {
                 completeTaskCloseClean(task);
-                cleanUpTaskProducer(task, taskCloseExceptions);
-                tasks.remove(task.id());
+                if (task.isActive()) {
```

Comment:
The old `cleanUpTaskProducer` contained this check, while the new `cleanUpTaskProducerAndRemoveTask` does not --  maybe we could move the check into the new method, but was not sure about it.

Reply:
IIUC the direction you want to head in (this is not the final PR, right? Lots of TODOs) is to improve type safety and ultimately have most of the methods in `Tasks` accept specific StreamTask/StandbyTask, vs just a generic "Task". Is that correct? In that case it feels correct to leave it here.
But if the ultimate goal is to make the TaskManager _more_ task-type agnostic, then maybe it should be in Tasks. How would you define the division of labor between TaskManager and Tasks?

Reply:
Your understanding is correct. TaskManager should know if it's dealing with an active or standby -- in fact, I want to remove some method like `addRecords()` from `Task` interface and move it to `StreamTask`, too.

And yes, this is just a first PR to keep the scope small.